File: base64_test.lua
local base64 = require "base64"

local map = "1qaz2wsx3edc4rfvbgt56yhnmju78ik,.lo90p;/'[-Q#$Y*&XM<>N~`ASRTOP()\""
local pad = "!"

local test = {}
test[#test+1] = "base64"
test[#test+1] = "o13=c./4230c-a]\'"
test[#test+1] = map
test[#test+1] = "$c*.<[}|?'\"t}{]-=IOUE4#"
test[#test+1] = "WkdSa1hISmtaR1JrWEhKY2JtUmtaR1JjYm1Oall3bz0="
test[#test+1] = "WjBHUjNYSDRKYWtaR1JYSEpjYm1Sa1pHUmNibU5qWXdvPTkw"
test[#test+1] = "bGVhc3VyZS4="
test[#test+1] = "28c930-acgoh.r>;']\\rt"
test[#test+1] = "cidsfaooewo412843c13-534209"
test[#test+1] = "MjhjOTMwLWFjZ29oLnI-OyddXHJ0"

-- API usage, beg:{
for i=1,#test do
	local rawstr = test[i]
	local encstr = base64.encode(rawstr)
	local decstr = base64.decode(encstr)
	if rawstr ~= decstr then
		local outstr = string.format("base64.decode(%q) => %q, expect %q", encstr, decstr, rawstr)
		print(outstr)
	end

	local rawstr = test[i]
	local encstr = base64.urlsafe_encode(rawstr)
	local decstr = base64.urlsafe_decode(encstr)
	if rawstr ~= decstr then
		local outstr = string.format("base64.urlsafe_decode(%q) => %q, expect %q", encstr, decstr, rawstr)
		print(outstr)
	end

	local schtab = {}
	schtab[1] = base64.scheme_stdmode()
	schtab[2] = base64.scheme_urlsafe()
	schtab[3] = base64.scheme_create(map, pad)

	for i=1,#schtab do
		local sch = schtab[i]
		local rawstr = test[i]
		local encstr = sch:encode(rawstr)
		local decstr = sch:decode(encstr)
		if rawstr ~= decstr then
			local outstr = string.format("sch:decode(%q) => %q, expect %q", encstr, decstr, rawstr);
			print("")
			print(outstr)
			if sch == schtab[1] then
				print("Using scheme_stdmode")
			end
			if sch == schtab[2] then
				print("Using scheme_urlsafe")
			end
			if sch == schtab[3] then
				print(string.format("Using map %q padding %q", map, pad))
			end
			print("")
		end
	end
end
-- API usage, end.}

print("TEST PASSED")

File: base64_test.c
#include "base64.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <stdlib.h>

struct pair {
	const char *rawstring;
	const char *stdencode;
	const char *urlencode;
};

#include "base64_test.h"

static void
TestStdmodeEncode(void) {
	for (const struct pair *p = pairs; p->rawstring != NULL; p++) {
		size_t len = strlen(p->rawstring);
		size_t dlen = base64_encode_length(len);
		char chs[dlen+1];
		base64_encode(p->rawstring, len, chs);
		chs[dlen] = '\0';
		if (strcmp(p->stdencode, chs) != 0) {
			fprintf(stderr, "stdmode_encode(%s) => %s, expect %s\n", p->rawstring, chs, p->stdencode);
			abort();
		}
	}
}

static void
TestStdmodeDecode(void) {
	for (const struct pair *p = pairs; p->stdencode != NULL; p++) {
		size_t len = strlen(p->stdencode);
		size_t mlen = base64_decode_length(len);
		char chs[mlen+1];
		size_t dlen;
		if (base64_decode(p->stdencode, len, chs, &dlen) < 0) {
			fprintf(stderr, "stdmode_decode(%s) fail! expect %s\n", p->stdencode, p->rawstring);
			abort();
		}
		chs[dlen] = '\0';
		if (strcmp(p->rawstring, chs) != 0) {
			fprintf(stderr, "stdmode_decode(%s) => %s, expect %s\n", p->stdencode, chs, p->rawstring);
			abort();
		}
	}
}

static void
TestUrlsafeEncode(void) {
	for (const struct pair *p = pairs; p->rawstring != NULL; p++) {
		size_t len = strlen(p->rawstring);
		size_t dlen = base64_encode_length(len);
		char chs[dlen+1];
		base64_urlsafe_encode(p->rawstring, len, chs);
		chs[dlen] = '\0';
		if (strcmp(p->urlencode, chs) != 0) {
			fprintf(stderr, "urlsafe_encode(%s) => %s, expect %s\n", p->rawstring, chs, p->urlencode);
			abort();
		}
	}
}

static void
TestUrlsafeDecode(void) {
	for (const struct pair *p = pairs; p->urlencode != NULL; p++) {
		size_t len = strlen(p->urlencode);
		size_t mlen = base64_decode_length(len);
		char chs[mlen+1];
		size_t dlen;
		long ret;
		if ((ret = base64_urlsafe_decode(p->urlencode, len, chs, &dlen)) < 0) {
			fprintf(stderr, "urlsafe_decode(%s) fail(%ld)! expect %s\n", p->urlencode, ret, p->rawstring);
			abort();
		}
		chs[dlen] = '\0';
		if (strcmp(p->rawstring, chs) != 0) {
			fprintf(stderr, "urlsafe_decode(%s) => %s, expect %s\n", p->urlencode, chs, p->rawstring);
			abort();
		}
	}
}

static void
TestScheme(void) {
	const char *map = "1qaz2wsx3edc4rfvbgt56yhnmju78ik,.lo90p;/'[-Q#$Y*&XM<>N~`ASRTOP()\"";
	int padding = '!';
	struct base64_scheme *sch = malloc(base64_scheme_length());
	base64_scheme_launch(sch, map, padding);
	const char *raw = "$c*.<[}|?'\"t}{]-=IOUE4#";
	size_t rawlen = strlen(raw);
	size_t enclen = base64_encode_length(rawlen);
	char enc[enclen+1];
	base64_scheme_encode(sch, raw, rawlen, enc);
	enc[enclen] = '\0';
	char dec[base64_decode_length(enclen)+1];
	size_t declen;
	base64_scheme_decode(sch, enc, enclen, dec, &declen);
	dec[declen] = '\0';
	if (strcmp(raw, dec) != 0) {
		fprintf(stderr, "MAP:\n%s\n\n",  map);
		fprintf(stderr, "raw:\n%s\nenc:\n%s\ndec:\n%s\n", raw, enc, dec);
		abort();
	}
}

int
main(void) {
	setvbuf(stdout, NULL, _IONBF, 0);
	setvbuf(stderr, NULL, _IONBF, 0);

	TestStdmodeEncode();
	TestStdmodeDecode();
	TestUrlsafeEncode();
	TestUrlsafeDecode();
	TestScheme();
	puts("TEST PASSED");
	return 0;
}
